There is a failing test:
```
[exist-expathrepo] [INFO] Running xquery.modules.expathrepo.ExpathRepoTests
Error: expathrepo] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.760 s <<< FAILURE! -- in xquery.modules.expathrepo.ExpathRepoTests
Error: expathrepo] [ERROR] xquery.modules.expathrepo.ExpathRepoTests -- Time elapsed: 0.760 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "org.exist.xquery.Expression.accept(org.exist.xquery.ExpressionVisitor)" because the return value of "org.exist.xquery.FilteredExpression.getExpression()" is null
	at org.exist.xquery.DefaultExpressionVisitor.visitFilteredExpr(DefaultExpressionVisitor.java:189)
	at org.exist.xquery.FilteredExpression.accept(FilteredExpression.java:224)
	at org.exist.xquery.DefaultExpressionVisitor.visitBuiltinFunction(DefaultExpressionVisitor.java:51)
```

--- a/exist-core/src/main/java/org/exist/xquery/DefaultExpressionVisitor.java
+++ b/exist-core/src/main/java/org/exist/xquery/DefaultExpressionVisitor.java
@@ -186,8 +186,9 @@ public class DefaultExpressionVisitor extends BasicExpressionVisitor {
 
     @Override
     public void visitFilteredExpr(FilteredExpression filtered) {
-        if (filtered.getExpression() != null) {
-            filtered.getExpression().accept(this);
+        @Nullable final Expression expr = filtered.getExpression();
+        if (expr != null) {
+            expr.accept(this);
         }
         for (final Predicate pred : filtered.getPredicates()) {
             pred.accept(this);

--- a/exist-core/src/main/java/org/exist/xquery/DefaultExpressionVisitor.java
+++ b/exist-core/src/main/java/org/exist/xquery/DefaultExpressionVisitor.java
@@ -186,7 +186,9 @@ public class DefaultExpressionVisitor extends BasicExpressionVisitor {
 
     @Override
     public void visitFilteredExpr(FilteredExpression filtered) {
-        filtered.getExpression().accept(this);
+        if (filtered.getExpression() != null) {
+            filtered.getExpression().accept(this);
+        }
         for (final Predicate pred : filtered.getPredicates()) {
             pred.accept(this);
         }

--- a/exist-core/src/main/java/org/exist/xquery/Optimizer.java
+++ b/exist-core/src/main/java/org/exist/xquery/Optimizer.java
@@ -259,6 +259,13 @@ public class Optimizer extends DefaultExpressionVisitor {
                 predicate = (Predicate) parent;
             }
 
+            // The predicate splitting optimization only applies when the
+            // predicate belongs to a LocationStep — not a FilteredExpression
+            // or other expression type that also carries predicates.
+            if (!(predicate.getParent() instanceof LocationStep)) {
+                return;
+            }
+
             if (LOG.isTraceEnabled()) {
                 LOG.trace("Rewriting boolean expression: {}", ExpressionDumper.dump(and));
             }

